### PR TITLE
feat(dashboard): add RBAC feature for Dashboard

### PR DESCRIPTION
**Pull request by lafirest**
Fixes [EMQX-10977](https://emqx.atlassian.net/browse/EMQX-10977)

### <samp>🤖 Generated by Copilot at 5c31c5c</samp>

This pull request adds the role management feature for the dashboard users, which allows specifying and checking the user role based on the RBAC permissions. It introduces a new application `emqx_dashboard_rbac`, which handles the RBAC logic and depends on the `emqx_dashboard` application. It also modifies several files in the `emqx_dashboard` application to support different user roles and RBAC checks depending on the release edition.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: changes/ee/feat-11610.en.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Implemented a preliminary Role-Based Access Control for the Dashboard.
+In this version, there are two predefined roles:
+- superuser
+  This role could access all resources.
+- viewer
+  This role only can access the `GET` resource.
```
Comment (wivwiv) [inline changes/ee/feat-11610.en.md:6]:
```suggestion
  This role can only view resources and data, corresponding to all GET requests in the REST API.
```

**Review by wivwiv — CHANGES_REQUESTED** [review]

File: apps/emqx_dashboard_rbac/src/emqx_dashboard_rbac.erl [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    ?ROLE_SUPERUSER;
+role(#?ADMIN{role = Role}) ->
+    Role;
+%% For compatibility
+role([]) ->
+    ?ROLE_SUPERUSER;
+role(#{role := Role}) ->
+    Role.
+
+legal_role(Role) ->
```
Comment (zmstone) [inline apps/emqx_dashboard_rbac/src/emqx_dashboard_rbac.erl:30]: valid_role

File: apps/emqx_dashboard/src/emqx_dashboard_api.erl [hunk 3]
```diff
@@ -203,7 +204,9 @@ field(version) ->
 field(old_pwd) ->
     {old_pwd, mk(binary(), #{desc => ?DESC(old_pwd)})};
 field(new_pwd) ->
-    {new_pwd, mk(binary(), #{desc => ?DESC(new_pwd)})}.
+    {new_pwd, mk(binary(), #{desc => ?DESC(new_pwd)})};
+field(role) ->
+    {role, mk(binary(), #{desc => ?DESC(role), example => ?ROLE_DEFAULT})}.
```
Comment (zmstone) [inline apps/emqx_dashboard/src/emqx_dashboard_api.erl:209]: we'll need to bump api version for this endpoint?
if the old rest-api client calls the same api to create a user, they would expect it to be superuser.

Comment (lafirest) [reply apps/emqx_dashboard/src/emqx_dashboard_api.erl:209]: no, this change is invisable for open source users, I will change the default role to the same as the old version

File: apps/emqx_dashboard/src/emqx_dashboard.erl [hunk 4]
```diff
@@ -205,13 +205,16 @@ authorize(Req) ->
         {basic, Username, Password} ->
             api_key_authorize(Req, Username, Password);
         {bearer, Token} ->
-            case emqx_dashboard_admin:verify_token(Token) of
+            case emqx_dashboard_admin:verify_token(Req, Token) of
```
Comment (JimMoen) [inline apps/emqx_dashboard/src/emqx_dashboard.erl:208]: maybe a better func name like: `emqx_dashboard_admin:verify_rbac_token/2` ?

File: apps/emqx_dashboard_rbac/BSL.txt [hunk 5]
```diff
@@ -0,0 +1,94 @@
+Business Source License 1.1
```
Comment (HJianBo) [inline apps/emqx_dashboard_rbac/BSL.txt:1]: Should it be placed in the lib-ee directory if it is a BSL application? Also, we should not pack a BSL application when doing releases.

**Review by HJianBo — APPROVED** [review]

**Pull request merged**